### PR TITLE
Adding segmented mean/min/max in TF and Pypark

### PR DESCRIPTION
**Pull request by anwoods894**
### Description

Changes include the addition of Spark and TF code for calculating list-wise statistics, segmented by a second input. This allows for relative features within subcategories of items in ranking/selection tasks, e.g. price relative to other 4 star hotels, price relative to other refundable rates.

The below checklists come from the docs page on adding new transformers [here](../docs/adding_transformer.md)
### Keras Layer Checklist
Verify that:
- [x] The new Keras layer extends [BaseLayer](../src/kamae/tensorflow/layers/base.py)
- [x] The `_call` method has been implemented in the new layer.
- [x] The `compatible_dtypes` property is defined in the new layer.
- [x] The new layer is decorated with `@tf.keras.utils.register_keras_serializable(package=kamae.__name__)`.
- [x] The new layer takes a `name`, `input_dtype`, and `output_dtype` as arguments to the constructor and that this is passed to the super constructor.
- [x] The Keras layer is serializable. I have implemented the `get_config` method.
- [x] There are unit tests of the new layer. 
- [x] There is a specific test of layer serialisation added [here](../tests/kamae/tensorflow/test_layer_serialisation.py).
- [x] The new layer is imported in the [__init__.py](../src/kamae/tensorflow/layers/__init__.py) file in the `layers` directory.

### Spark Transformer/Estimator Checklist
Verify that:
- [x] The new Spark Transformer extends [BaseTransformer](../src/kamae/spark/transformers/base.py).
- [x] If the new transform needs a fit method, a Spark Estimator has been implemented that extends [BaseEstimator](../src/kamae/spark/estimators/base.py).
- [x] The instructions in the above docs page have been followed for the `__init__` and `setParams` methods.
- [x] The transformer uses one of the input/output mixin classes from [base.py](../src/kamae/spark/params/base.py).
- [x] If the new transformer requires more parameters that would need to be serialised to the Spark ML pipeline, there is a implemented parameter class by extending the `Params` class [here](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.ml.param.Params.html).
- [x] The `compatible_dtypes` property has been implemented to specify the input/output data types that my transformer/estimator supports.
- [x] A Keras subclassed layer is returned in the transformer's `get_tf_layer` method.
- [x] There are unit tests of the new transform. In particular, there are parity tests between the Spark and Keras implementations.
- [x] The new transformer/estimator is imported in the [__init__.py](../src/kamae/spark/transformers/__init__.py) file in the `transformers`/`estimators` directory.

Finally, please verify that:
- [x] There is a new entry (alphabetical order) in the README table describing the new layer/transformer

### CODE REVIEW

File: src/kamae/tensorflow/layers/segment_max.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            return tf.reshape(gathered, tf.shape(v))
+
+        values = inputs[0]
+        segments = inputs[1]
+
+        return tf.map_fn(
+            lambda x: per_batch_max(x[0], x[1]),
+            (values, segments),
+            fn_output_signature=tf.TensorSpec(shape=values.shape[1:], dtype=tf.float32),
+        )
```
Comment (georyetti) [inline src/kamae/tensorflow/layers/segment_max.py:113]: We have a util for using map_fn with an axis parameter [here](https://github.com/ExpediaGroup/kamae/blob/9781bdc315db3d97b2581470300e1ebd2249bb08/src/kamae/tensorflow/utils/transform_utils.py#L22) can we use it instead? That way we can add an axis param to this layer like the other listwise ops (e.g. [here](https://github.com/ExpediaGroup/kamae/blob/9781bdc315db3d97b2581470300e1ebd2249bb08/src/kamae/tensorflow/layers/list_max.py#L53)). We don't set the axis from Spark -> Keras to keep the default, but if the user is using the layer without the transformer, we should allow the flexibility.

Comment (georyetti) [reply src/kamae/tensorflow/layers/segment_max.py:113]: Ah in fact you do have an axis param, but its not used currently.

File: src/kamae/tensorflow/layers/segment_mean.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        ]
+
+    @allow_single_or_multiple_tensor_input
+    def _call(self, inputs: Iterable[Tensor], **kwargs) -> Tensor:
+        """
+        Calculate the mean of the first input,
+        segmented by the values of the second input.
+
+        :param inputs: The iterable tensors for the features.
+        :returns: Thew new tensor result column.
```
Comment (georyetti) [inline src/kamae/tensorflow/layers/segment_mean.py:88]: Typo "Thew"

**Review by georyetti — CHANGES_REQUESTED** [review]
- Axis is currently unused, let's use the map fn util as described in the comments
- Some minor typos
- Potential renaming to make it clear its a listwise op

File: src/kamae/spark/transformers/segment_max.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    :param inputCols: Input column name.
+    - The first is the value column, on which to calculate the maximum.
+    - The second is the segment column, based on which to segment the items.
+    :param outputCol: Name of output col.
+    :param inputDtype: Data Type of input.
+    :param outputDtype: Data Type of output.
+    :param layerName: The name of the transformer, which typically
+    should be the name of the produced feature.
+    :param queryIdCol: Name of column to denoting a single query. It is required.
+    """
```
Comment (jamesdshinner) [inline src/kamae/spark/transformers/segment_max.py:58]: I'm not very clear on what this last param doc is saying

File: src/kamae/tensorflow/layers/segment_mean.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        def per_batch_mean(v, s):
+            flat_v = tf.reshape(v, [-1])
+            flat_s = tf.reshape(s, [-1])
+            unique_segments, segment_indices = tf.unique(flat_s)
+            num_segments = tf.size(unique_segments)
+
+            # Cast
+            flat_v = tf.cast(flat_v, tf.float32)
+
+            min_vals = tf.math.unsorted_segment_mean(
```
Comment (jamesdshinner) [inline src/kamae/tensorflow/layers/segment_mean.py:100]: This is called min_vals but we're on the mean layer

**Review by jamesdshinner — CHANGES_REQUESTED** [review]

File: src/kamae/spark/transformers/segment_max.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+)
+
+from kamae.spark.params import ListwiseStatisticsParams, MultiInputSingleOutputParams
+from kamae.spark.utils import check_listwise_columns, get_listwise_condition_and_window
+from kamae.tensorflow.layers import SegmentMaxLayer
+
+from .base import BaseTransformer
+
+
+class SegmentMaxTransformer(
```
[the remark this answers is not in the document]
Comment (ddonghi) [reply src/kamae/spark/transformers/segment_max.py:39]: I agree. It should be easy to spot listwise operations, since they are fundamentally different, with more side/implicit effects (eg outliers impact in listwise mean calculation).

File: src/kamae/spark/transformers/segment_max.py (hunk 3 above)
Comment (ddonghi) [reply src/kamae/spark/transformers/segment_max.py:58]: For listwise operations, in Spark, we need to be able to define what is the grouping column. Normally it is "search_id" or "qid", but can be something else.

File: src/kamae/tensorflow/layers/segment_max.py (hunk 1 above)
Comment (ddonghi) [reply src/kamae/tensorflow/layers/segment_max.py:113]: In general, can we think about any way to avoid at all "map_fn"? It is slow! :)

**Comment by ddonghi** [comment]
> Happy to add a segment to the listwise ops, which defaults to `None` to maintain backwards compat

Agree. If you use a "segment" arg, it may be the transformer will not propagate the column to read from the dataset, since not in inputCol(s). If this is true (please check), you can just assume the segment is in inputCols when an arg like `with_segment` is set to True. Example, the last inputCols is the segment if `with_segment=true`

What do you think?

**Comment by anwoods894** [comment]
Thanks for the comments all! Agree, will look into combining this into the existing list ops.

**Comment by anwoods894** [comment]
Changes include:
- Addition of segmented functionality to the existing listwise transformers 
    - `with_segment` decides the role of the second input, defaulting to `False`.
- Adapted `map_fn_w_axis` to optionally take two tensors as input
- Tried a few things to avoid using `map_fn` at all but could not find a nice way of handling some edge cases (same segment IDs repeated over multiple batches/queries, for example).

File: src/kamae/spark/transformers/list_max.py [hunk 6]
```diff
@@ -116,10 +119,7 @@ def compatible_dtypes(self) -> Optional[List[DataType]]:
         return [
             FloatType(),
             DoubleType(),
-            ByteType(),
-            ShortType(),
-            IntegerType(),
-            LongType(),
+            StringType(),
```
Comment (georyetti) [inline src/kamae/spark/transformers/list_max.py:122]: Does this transformer work with strings? How does max work in this scenario?

File: src/kamae/spark/transformers/list_max.py [hunk 7]
```diff
@@ -50,31 +42,40 @@ class ListMaxTransformer(
...
+    by the value of withSegment as described above.
     :param outputCol: Name of output col.
     :param inputDtype: Data Type of input.
     :param outputDtype: Data Type of output.
     :param layerName: The name of the transformer, which typically
     should be the name of the produced feature.
     :param queryIdCol: Name of column to aggregate upon. It is required.
-    :param topN: Filter for limiting the items to calculate the statistics.
+    :param topN: Filter for limiting the items to calculate the statistics. Not used when withSegment = True.
     :param sortOrder: Option of 'asc' or 'desc' which defines order
```
Comment (georyetti) [inline src/kamae/spark/transformers/list_max.py:75]: Is there any reason why we can't have segment, sort and filtering based on topN together?

File: src/kamae/spark/transformers/list_mean.py [hunk 8]
```diff
@@ -108,6 +128,11 @@ def compatible_dtypes(self) -> Optional[List[DataType]]:
         return [
             FloatType(),
             DoubleType(),
+            ByteType(),
+            ShortType(),
+            IntegerType(),
+            LongType(),
```
Comment (georyetti) [inline src/kamae/spark/transformers/list_mean.py:134]: We removed these for max, do they work for mean?

File: src/kamae/tensorflow/utils/list_utils.py [hunk 9]
```diff
@@ -82,3 +82,24 @@ def listify_tensors(x: Union[tf.Tensor, np.ndarray, List[Any]]) -> List[Any]:
     if isinstance(x, np.ndarray):
         x = x.tolist()
     return x
+
+
+def segmented_operation(values: List, fn: Callable) -> Tensor:
```
Comment (georyetti) [inline src/kamae/tensorflow/utils/list_utils.py:87]: Can we typehint this as `List[Tensor]` pls

**Review by georyetti — CHANGES_REQUESTED** [review]
Couple of comments:
1. We added string support to some of them, not sure how this works with numeric statistics. If it does work, can we add some tests please? 
2. We have more listwise stats, e.g. median & stddev, can we add segmentation to these too? If you want it can be another PR, but let's make sure they don't get left behind.
3. Some statistics had integer support removed, others didn't, is that expected?
4. Can we have segmenting & sorting/filtering? E.g if we have two inputs and withSegment=True, assume second column is segment and no sorting. If withSegment=False with two inputs then assume sorting without segmenting. If we have 3 inputs then withSegment must be True and assume second column is sorting and third column is segmenting...?

File: src/kamae/spark/transformers/list_max.py (hunk 6 above)
Comment (georyetti) [reply src/kamae/spark/transformers/list_max.py:122]: Ah yeah that's frustrating, ok happy to keep as is. In the future the compatible dtypes may need a refactor as some other layers have the same issue

File: src/kamae/spark/transformers/list_max.py (hunk 7 above)
Comment (georyetti) [reply src/kamae/spark/transformers/list_max.py:75]: Agreed to do this in a new PR

**Review by georyetti — DISMISSED** [review]

File: src/kamae/spark/params/shared.py [hunk 10]
```diff
@@ -510,6 +518,24 @@ def getSortOrder(self) -> str:
         """
         return self.getOrDefault(self.sortOrder)
 
+    def setWithSegment(self, value: bool) -> "ListwiseStatisticsParams":
+        """
+        Sets the query id parameter.
```
Comment (jamesdshinner) [inline src/kamae/spark/params/shared.py:523]: Incorrect docstring from copy-paste

File: src/kamae/spark/transformers/list_max.py [hunk 11]
```diff
@@ -50,31 +42,40 @@ class ListMaxTransformer(
...
-    By using the topN items to calculate the statistics, we can better approximate
-    the real statistics in production. It should be used a large enough topN to get a
-    good approximation of the statistics, and an important feature to sort on, such as
-    item's production.
+        - If withSegment = False: the transformer calculates the maximum of the first column
+        with the same query id column value, based on second column's topN items.
+        When using the second input as sorting column, topN must be provided.
+        By using the topN items to calculate the statistics, we can better approximate
+        the real statistics in production. It should be used a large enough topN to get a
+        good approximation of the statistics, and an important feature to sort on, such as
```
Comment (jamesdshinner) [inline src/kamae/spark/transformers/list_max.py:57]: Can we re-word "It should be used a large enough topN"?

File: src/kamae/tensorflow/layers/list_max.py [hunk 12]
```diff
@@ -92,63 +107,67 @@ def compatible_dtypes(self) -> Optional[List[tf.dtypes.DType]]:
...
+                    val_tensor=val_tensor,
+                    axis=self.axis,
+                    sort_tensor=sort_tensor,
+                    sort_order=self.sort_order,
+                    top_n=self.top_n,
+                )
         else:
-            filtered_tensor = val_tensor
+            if self.with_segment:
+                raise ValueError("with_segment set to True, expected two inputs.")
```
Comment (jamesdshinner) [inline src/kamae/tensorflow/layers/list_max.py:143]: Are there more cases we should error other than `len(inputs) == 2 and self.with_segment`?

**Review by jamesdshinner — CHANGES_REQUESTED** [review]
Some minor feedback

File: src/kamae/tensorflow/layers/list_max.py (hunk 12 above)
Comment (anwoods894) [reply src/kamae/tensorflow/layers/list_max.py:143]: I don't think so. If you specify `top_n` or `sort_order` without a second input, they just don't get used. The code that does the statistic calculation doesn't change, since for sorting we are just modifying `val_tensor`. 

In the case where `self.with_segment = True` we use a different fn, and if not caught here we'll get an error from line 155.

I can think of the case where the user passes in >2 inputs, which should probably raise some error or warning, but since I am hoping to expand this functionality to take 3 inputs in my next PR, handling that case can come in that PR?

File: src/kamae/spark/params/shared.py (hunk 10 above)
Comment (anwoods894) [reply src/kamae/spark/params/shared.py]: Done!
